// Code generated by ent, DO NOT EDIT.

package ent

import (
	"context"
	"resume-builder-backend/ent/predicate"
	"resume-builder-backend/ent/resume"

	"entgo.io/ent/dialect/sql"
	"entgo.io/ent/dialect/sql/sqlgraph"
	"entgo.io/ent/schema/field"
)

// ResumeDelete is the builder for deleting a Resume entity.
type ResumeDelete struct {
	config
	hooks    []Hook
	mutation *ResumeMutation
}

// Where appends a list predicates to the ResumeDelete builder.
func (_d *ResumeDelete) Where(ps ...predicate.Resume) *ResumeDelete {
	_d.mutation.Where(ps...)
	return _d
}

// Exec executes the deletion query and returns how many vertices were deleted.
func (_d *ResumeDelete) Exec(ctx context.Context) (int, error) {
	return withHooks(ctx, _d.sqlExec, _d.mutation, _d.hooks)
}

// ExecX is like Exec, but panics if an error occurs.
func (_d *ResumeDelete) ExecX(ctx context.Context) int {
	n, err := _d.Exec(ctx)
	if err != nil {
		panic(err)
	}
	return n
}

func (_d *ResumeDelete) sqlExec(ctx context.Context) (int, error) {
	_spec := sqlgraph.NewDeleteSpec(resume.Table, sqlgraph.NewFieldSpec(resume.FieldID, field.TypeInt))
	if ps := _d.mutation.predicates; len(ps) > 0 {
		_spec.Predicate = func(selector *sql.Selector) {
			for i := range ps {
				ps[i](selector)
			}
		}
	}
	affected, err := sqlgraph.DeleteNodes(ctx, _d.driver, _spec)
	if err != nil && sqlgraph.IsConstraintError(err) {
		err = &ConstraintError{msg: err.Error(), wrap: err}
	}
	_d.mutation.done = true
	return affected, err
}

// ResumeDeleteOne is the builder for deleting a single Resume entity.
type ResumeDeleteOne struct {
	_d *ResumeDelete
}

// Where appends a list predicates to the ResumeDelete builder.
func (_d *ResumeDeleteOne) Where(ps ...predicate.Resume) *ResumeDeleteOne {
	_d._d.mutation.Where(ps...)
	return _d
}

// Exec executes the deletion query.
func (_d *ResumeDeleteOne) Exec(ctx context.Context) error {
	n, err := _d._d.Exec(ctx)
	switch {
	case err != nil:
		return err
	case n == 0:
		return &NotFoundError{resume.Label}
	default:
		return nil
	}
}

// ExecX is like Exec, but panics if an error occurs.
func (_d *ResumeDeleteOne) ExecX(ctx context.Context) {
	if err := _d.Exec(ctx); err != nil {
		panic(err)
	}
}
